datadog_lambda/dsm.py

```diff
@@ -24,15 +36,88 @@ def _dsm_set_sqs_context(event):
 
     for record in records:
         try:
-            queue_arn = record.get("eventSourceARN", "")
-
-            contextjson = get_datastreams_context(record)
-            payload_size = calculate_sqs_payload_size(record)
+            arn = arn_extractor(record)
+            context_json = _get_dsm_context_from_lambda(record)
+            payload_size = payload_size_calculator(record, context_json)
 
-            ctx = DsmPathwayCodec.decode(contextjson, processor)
+            ctx = DsmPathwayCodec.decode(context_json, processor)
             ctx.set_checkpoint(
-                ["direction:in", f"topic:{queue_arn}", "type:sqs"],
+                ["direction:in", f"topic:{arn}", f"type:{service_type}"],
                 payload_size=payload_size,
             )
         except Exception as e:
             logger.error(format_err_with_traceback(e))
+
+
+def _dsm_set_sqs_context(event):
+    from ddtrace.internal.datastreams.botocore import calculate_sqs_payload_size
+
+    def sqs_payload_calculator(record, context_json):
+        return calculate_sqs_payload_size(record)
+
+    def sqs_arn_extractor(record):
+        return record.get("eventSourceARN", "")
+
+    _dsm_set_context_helper(event, "sqs", sqs_arn_extractor, sqs_payload_calculator)
+
+
+def _get_dsm_context_from_lambda(message):
+    """
+    Lambda-specific message formats:
+        - message.messageAttributes._datadog.stringValue (SQS -> lambda)
+        - message.Sns.MessageAttributes._datadog.Value.decode() (SNS -> lambda)
+        - message.messageAttributes._datadog.binaryValue.decode() (SNS -> SQS -> lambda, raw)
+        - message.body.MessageAttributes._datadog.Value.decode() (SNS -> SQS -> lambda)
+        - message.kinesis.data.decode()._datadog (Kinesis -> lambda)
+    """
+    context_json = None
+    message_body = message
+
+    if "kinesis" in message:
```

Comment:
Can we simplify this PR. Right now main doesn't even have kinesis support, so it's impossible for me to review this effectively. Start first with a single PR that does the change you're looking for but _just_ for SQS.